networks/1_correlation_analysis.py

```diff
@@ -59,143 +59,177 @@
+h2.fig.suptitle(
+    f"Correlation of PA14 genes (SPELL corrected using {num_singular_values} vectors)",
+    y=1.05,
 )
-pa14_local_dist = pa14_local_dist.stack().reset_index()
-pa14_local_dist.columns = ["Row", "Column", "Value"]
 
-pa14_local_dist.head(10)
+# Save
+pa14_spell_filename = os.path.join(paths.LOCAL_DATA_DIR, "pa14_spell_clustermap.png")
+h2.savefig(pa14_spell_filename, dpi=300)
 # -
 
-f4 = sns.displot(pa14_local_dist["Value"])
-plt.title("Distribution of pairwise distances for PA14 genes")
+# Save SPELL correlation matrices
+pao1_spell_mat_filename = os.path.join(paths.LOCAL_DATA_DIR, "pao1_spell_mat.tsv")
+pa14_spell_mat_filename = os.path.join(paths.LOCAL_DATA_DIR, "pa14_spell_mat.tsv")
+pao1_corr_spell.to_csv(pao1_spell_mat_filename, sep="\t")
+pa14_corr_spell.to_csv(pa14_spell_mat_filename, sep="\t")
 
-# ## Plot correlation
-#
-# We will plot a heatmap and umap of the correlations to identify clusters, which should help to inform the parameters for hierarchal clustering - i.e. how many clusters can we expect?
+# ## log transform + SPELL + correlation
+
+# log transform data
+pao1_compendium_log10 = np.log10(1 + pao1_compendium_T)
+pa14_compendium_log10 = np.log10(1 + pa14_compendium_T)
+
+# Apply SVD
+pao1_U, pao1_s, pao1_Vh = np.linalg.svd(pao1_compendium_log10, full_matrices=False)
+pa14_U, pa14_s, pa14_Vh = np.linalg.svd(pa14_compendium_log10, full_matrices=False)
+
+print(pao1_compendium_T.shape)
+print(pao1_U.shape, pao1_s.shape, pao1_Vh.shape)
+
+print(pa14_compendium_T.shape)
+print(pa14_U.shape, pa14_s.shape, pa14_Vh.shape)
+
+# Convert ndarray to df to use corr()
+pao1_U_df = pd.DataFrame(data=pao1_U, index=pao1_compendium_T.index)
+pa14_U_df = pd.DataFrame(data=pa14_U, index=pa14_compendium_T.index)
+
+# +
+# Correlation of U
+# Since `corr()` computes pairwise correlation of columns we need to invert U
+
+pao1_corr_log_spell = pao1_U_df.iloc[:, :num_singular_values_log].T.corr()
+pa14_corr_log_spell = pa14_U_df.iloc[:, :num_singular_values_log].T.corr()
 
 # +
 # Plot heatmap
-plt.figure(figsize=(20, 20))
-h1 = sns.clustermap(pao1_corr.abs(), cmap="viridis")
-h1.fig.suptitle(f"Correlation of PAO1 genes using threshold={corr_threshold}")
+h1a = sns.clustermap(pao1_corr_log_spell.abs(), cmap="viridis", figsize=(20, 20))
+h1a.fig.suptitle(
+    f"log transform + SPELL corrected using {num_singular_values} vectors (PAO1)",
+    y=1.05,
+)
 
 # Save
-pao1_clustermap_filename = os.path.join(
-    paths.LOCAL_DATA_DIR, f"pao1_corr_{corr_threshold}_clustermap.png"
+pao1_log_spell_filename = os.path.join(
+    paths.LOCAL_DATA_DIR, "pao1_log_spell_clustermap.png"
 )
-h1.savefig(pao1_clustermap_filename, dpi=300)
+h1a.savefig(pao1_log_spell_filename, dpi=300)
 
 # +
-plt.figure(figsize=(20, 20))
-h2 = sns.clustermap(pa14_corr.abs(), cmap="viridis")
-h2.fig.suptitle(f"Correlation of PA14 genes using threshold={corr_threshold}")
+h2a = sns.clustermap(pa14_corr_log_spell.abs(), cmap="viridis", figsize=(20, 20))
```

Comment:
It's interesting that the log transform really seems to be a necessary step in getting rid of the block diagonal structure (i.e. the large, highly connected network components).

Do you think the large communities were due to correlations in genes with very low expression? So the log transformation may have smoothed out these low expression values? (Just a guess)

Reply:
So based on the math, I believe these large communities are due to those genes that are in the right tail of the distribution - these are genes that will cause the covariance to be very high just because of how far it deviates from the mean.